// postcss-calc-ast-parser/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AST = require("./types/ast");
exports.AST = AST;
const css_calc_1 = require("./css-calc");
exports.Parser = css_calc_1.Parser;
exports.Tokenizer = css_calc_1.Tokenizer;
exports.Stringifier = css_calc_1.Stringifier;
exports.getResolvedType = css_calc_1.getResolvedType;
exports.reduceExpression = css_calc_1.reduceExpression;
exports.mathExpr = css_calc_1.newMathExpression;
function parse(code, options) {
    const tokenizer = new css_calc_1.Tokenizer(code, options);
    return new css_calc_1.Parser(tokenizer, options).parse();
}
exports.parse = parse;
function stringify(node, options) {
    const stringifier = new css_calc_1.Stringifier(options);
    return stringifier.stringify(node);
}
exports.stringify = stringify;
exports.default = {
    parse,
    stringify,
    getResolvedType: css_calc_1.getResolvedType,
    reduceExpression: css_calc_1.reduceExpression,
    mathExpr: css_calc_1.newMathExpression,
    Parser: css_calc_1.Parser,
    Tokenizer: css_calc_1.Tokenizer,
    Stringifier: css_calc_1.Stringifier,
    AST,
};
