// @tokens-studio/types/dist/constants/Properties.cjs
'use strict';

exports.Properties = void 0;
(function (Properties) {
    Properties["sizing"] = "sizing";
    Properties["height"] = "height";
    Properties["width"] = "width";
    Properties["spacing"] = "spacing";
    Properties["verticalPadding"] = "verticalPadding";
    Properties["horizontalPadding"] = "horizontalPadding";
    Properties["paddingTop"] = "paddingTop";
    Properties["paddingRight"] = "paddingRight";
    Properties["paddingBottom"] = "paddingBottom";
    Properties["paddingLeft"] = "paddingLeft";
    Properties["itemSpacing"] = "itemSpacing";
    Properties["fill"] = "fill";
    Properties["backgroundBlur"] = "backgroundBlur";
    Properties["border"] = "border";
    Properties["borderTop"] = "borderTop";
    Properties["borderRight"] = "borderRight";
    Properties["borderBottom"] = "borderBottom";
    Properties["borderLeft"] = "borderLeft";
    Properties["borderColor"] = "borderColor";
    Properties["borderRadius"] = "borderRadius";
    Properties["borderRadiusTopLeft"] = "borderRadiusTopLeft";
    Properties["borderRadiusTopRight"] = "borderRadiusTopRight";
    Properties["borderRadiusBottomRight"] = "borderRadiusBottomRight";
    Properties["borderRadiusBottomLeft"] = "borderRadiusBottomLeft";
    Properties["borderWidth"] = "borderWidth";
    Properties["borderWidthTop"] = "borderWidthTop";
    Properties["borderWidthRight"] = "borderWidthRight";
    Properties["borderWidthBottom"] = "borderWidthBottom";
    Properties["borderWidthLeft"] = "borderWidthLeft";
    Properties["boxShadow"] = "boxShadow";
    Properties["opacity"] = "opacity";
    Properties["fontFamilies"] = "fontFamilies";
    Properties["fontWeights"] = "fontWeights";
    Properties["fontSizes"] = "fontSizes";
    Properties["lineHeights"] = "lineHeights";
    Properties["typography"] = "typography";
    Properties["composition"] = "composition";
    Properties["letterSpacing"] = "letterSpacing";
    Properties["paragraphSpacing"] = "paragraphSpacing";
    Properties["textCase"] = "textCase";
    Properties["dimension"] = "dimension";
    Properties["textDecoration"] = "textDecoration";
    Properties["asset"] = "asset";
    Properties["tokenValue"] = "tokenValue";
    Properties["value"] = "value";
    Properties["tokenName"] = "tokenName";
    Properties["description"] = "description";
})(exports.Properties || (exports.Properties = {}));
