// @tokens-studio/types/dist/constants/TokenTypes.js
var TokenTypes;
(function (TokenTypes) {
    TokenTypes["OTHER"] = "other";
    TokenTypes["COLOR"] = "color";
    TokenTypes["BORDER_RADIUS"] = "borderRadius";
    TokenTypes["SIZING"] = "sizing";
    TokenTypes["SPACING"] = "spacing";
    TokenTypes["TEXT"] = "text";
    TokenTypes["TYPOGRAPHY"] = "typography";
    TokenTypes["OPACITY"] = "opacity";
    TokenTypes["BORDER_WIDTH"] = "borderWidth";
    TokenTypes["BOX_SHADOW"] = "boxShadow";
    TokenTypes["FONT_FAMILIES"] = "fontFamilies";
    TokenTypes["FONT_WEIGHTS"] = "fontWeights";
    TokenTypes["LINE_HEIGHTS"] = "lineHeights";
    TokenTypes["FONT_SIZES"] = "fontSizes";
    TokenTypes["LETTER_SPACING"] = "letterSpacing";
    TokenTypes["PARAGRAPH_SPACING"] = "paragraphSpacing";
    TokenTypes["PARAGRAPH_INDENT"] = "paragraphIndent";
    TokenTypes["TEXT_DECORATION"] = "textDecoration";
    TokenTypes["TEXT_CASE"] = "textCase";
    TokenTypes["COMPOSITION"] = "composition";
    TokenTypes["DIMENSION"] = "dimension";
    TokenTypes["BORDER"] = "border";
    TokenTypes["ASSET"] = "asset";
})(TokenTypes || (TokenTypes = {}));

export { TokenTypes };
