// @tokens-studio/sd-transforms/dist/index.cjs
'use strict';

var registerTransforms = require('./registerTransforms.cjs');
var expandComposites = require('./parsers/expand-composites.cjs');
var excludeParentKeys = require('./parsers/exclude-parent-keys.cjs');
var mapDescriptionToComment = require('./mapDescriptionToComment.cjs');
var checkAndEvaluateMath = require('./checkAndEvaluateMath.cjs');
var transformDimension = require('./transformDimension.cjs');
var transformFontWeights = require('./transformFontWeights.cjs');
var transformColorModifiers = require('./color-modifiers/transformColorModifiers.cjs');
var transformLineHeight = require('./transformLineHeight.cjs');
var transformShadow = require('./css/transformShadow.cjs');
var transformBorder = require('./css/transformBorder.cjs');
var transformTypography = require('./css/transformTypography.cjs');
var transformHEXRGBa = require('./css/transformHEXRGBa.cjs');
var transformLetterSpacing = require('./css/transformLetterSpacing.cjs');
var transformTypography$1 = require('./compose/transformTypography.cjs');



exports.registerTransforms = registerTransforms.registerTransforms;
exports.transforms = registerTransforms.transforms;
exports.expandComposites = expandComposites.expandComposites;
exports.excludeParentKeys = excludeParentKeys.excludeParentKeys;
exports.mapDescriptionToComment = mapDescriptionToComment.mapDescriptionToComment;
exports.checkAndEvaluateMath = checkAndEvaluateMath.checkAndEvaluateMath;
exports.transformDimension = transformDimension.transformDimension;
exports.transformFontWeights = transformFontWeights.transformFontWeights;
exports.transformColorModifiers = transformColorModifiers.transformColorModifiers;
exports.transformLineHeight = transformLineHeight.transformLineHeight;
exports.transformShadowForCSS = transformShadow.transformShadowForCSS;
exports.transformBorderForCSS = transformBorder.transformBorderForCSS;
exports.transformTypographyForCSS = transformTypography.transformTypographyForCSS;
exports.transformHEXRGBaForCSS = transformHEXRGBa.transformHEXRGBaForCSS;
exports.transformLetterSpacingForCSS = transformLetterSpacing.transformLetterSpacingForCSS;
exports.transformTypographyForCompose = transformTypography$1.transformTypographyForCompose;
